30053 fix log rollover timing

riyaaroraa opened the pull request:
SOE/OSX_15_SEQUOIA_ARM64_M2_SEMERU_OPEN_21.yml

fixes #30053

## Changes Made

- Increased wait time padding from 1000ms to 5000ms in TimeBasedLogRolloverTest
- Added server ready checks (waitForStringInLog("CWWKF0011I"))
- Added debug logging of directory contents before assertions
- Improved log file existence verification

## Testing

Verified locally with:

./gradlew :com.ibm.ws.logging_2_fat:buildandrun -Dtest.single=TimeBasedLogRolloverTest#testTimedRolloverEnv

## Checklist

I have considered the risk of behavior change or other zero migration impact
This PR fixes an Issue (N/A - test stability improvement)

## Notes for Reviewers

This addresses intermittent test failures where rolled log files weren't found due to:

- Tight timing constraints (fixed by increasing padding)
- Missing server-ready verification
- Insufficient debug info

riyaaroraa committed:
Fix time-based log rotation test timing issues - increased wait time and added debug logging

riyaaroraa committed:
Retrigger CI: Add debug comment

riyaaroraa commented:
#build

riyaaroraa commented:
!build

riyaaroraa commented:
!build

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=47c882e9-4994-4bf7-b344-d9aed8089b5f) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=4c5cedb3-c778-42ca-b615-5391a6e51548) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa committed:
Fix timing race conditions and server configuration issues

- Increased FILE_WAIT_SECONDS_PADDING from 1000ms to 5000ms
- Added CWWKO0221E to error ignore lists in test cleanup
- Added timedexit-1.0 feature to all server configurations
- Fixed JSP feature configuration for web app deployment

All 10 tests now passing locally in FULL mode

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=9d5a5433-b6e4-4dfa-a116-97b0d2deb794) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa committed:
Remove server ready wait that caused timing drift

- Removed waitForStringInLog(CWWKF0011I) from checkForRolledLogsAtTime
- This wait was causing 4-minute timeout and outdated rollover calculations
- Server configs are now correct so extra wait is unnecessary

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=ca2b694f-6238-43e5-9f43-8b568f4e7dcd) - :hourglass_flowing_sand:InProgress)

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/FfdcCleanupTest.java:
```diff
@@ -57,19 +57,32 @@ public static void initialSetup() throws Exception {
 
     }
 
+    /*
```
Should remove the commented out setup() method, to avoid confusion.

riyaaroraa replied:
Removed

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/FfdcCleanupTest.java:
```diff
@@ -57,19 +57,32 @@ public static void initialSetup() throws Exception {
 
```
Update the copyright year in the header, it should be `2022, 2025` instead of `2022, 2023`. Every time, we make changes to a file, we always update the copyright header year, to the current year. The format is `<YEAR OF FILE CREATED> , <YEAR OF LAST MODIFIED>`

riyaaroraa replied:
Fixed - updated copyright to 2022, 2025

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/FfdcCleanupTest.java:
```diff
@@ -57,19 +57,32 @@ public static void initialSetup() throws Exception {
             // Restore the original server configuration, before starting the server for each test case.
             server.restoreServerConfiguration();
-            server.startServer();
+            server.startServer(); // ← CHANGE THIS LINE
```
Remove the `CHANGE THIS LINE` comment, if its not needed.

riyaaroraa replied:
Removed

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/FfdcCleanupTest.java:
```diff
@@ -57,19 +57,32 @@ public static void initialSetup() throws Exception {
     public void cleanUp() throws Exception {
         if (server != null && server.isStarted()) {
-            server.stopServer("com.ibm.ws.logging.fat.ffdc.servlet", "ArithmeticException", "SRVE0777E");
+            server.stopServer("com.ibm.ws.logging.fat.ffdc.servlet", "ArithmeticException", "SRVE0777E", "CWWKO0221E");
```
The `CWWKO0221E` message notifies us if the `CWWKO0221E: TCP Channel defaultHttpEndpoint-ssl initialization did not succeed.` which might be useful for some cases, to help debug. I don't think we should put that messageID as expected error, since its not always expected.

riyaaroraa replied:
Removed CWWKO0221E as suggested

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -6,6 +6,7 @@
  * http://www.eclipse.org/legal/epl-2.0/
```
Update copyright year.

riyaaroraa replied:
Fixed - updated copyright to 2022, 2025

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -6,6 +6,7 @@
  * http://www.eclipse.org/legal/epl-2.0/
  *
  * SPDX-License-Identifier: EPL-2.0
+ * retrigger
```
We shouldn't modify the copyright header, with other information, as the copyright checker might fail. Please remove the "retrigger" wording.

riyaaroraa replied:
Removed

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -98,14 +101,31 @@ public static void initialSetup() throws Exception {
         ShrinkHelper.defaultDropinApp(server_xml, "quicklogtest", "com.ibm.ws.logging.fat.quick.log.test");
     }
 
+    /*
```
Remove comment, as it might be confusing.

riyaaroraa replied:
Removed

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -98,14 +101,31 @@ public static void initialSetup() throws Exception {
+     */
+
     public void setUp(LibertyServer server, String method) throws Exception {
+        LOG.info("Server starting at: " + new Date());
```
Suggest to move this line right before `server.startServer()`, so the time will be as accurate as possible.

riyaaroraa replied:
Done - Made the change as suggested

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -98,14 +101,31 @@ public static void initialSetup() throws Exception {
             // Restore the original server configuration, before starting the server for each test case.
             server.restoreServerConfiguration();
-            server.startServer();
+            server.startServer(); // ← CHANGE THIS LINE
```
Remove the comment for cleanliness.

riyaaroraa replied:
Done

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -124,7 +144,7 @@ private static void waitForBeginningOfMinute() {
         if (serverInUse != null && serverInUse.isStarted()) {
             serverInUse.stopServer("com.ibm.ws.logging.fat.ffdc.servlet.FFDCServlet.doGet", "ArithmeticException",
-                                   "CWWKG0081E", "CWWKG0083W", "TRAS3015W", "TRAS3013W");
+                                   "CWWKG0081E", "CWWKG0083W", "TRAS3015W", "TRAS3013W", "CWWKO0221E");
```
Same as above.

riyaaroraa replied:
Removed CWWKO0221E as suggested

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -404,33 +424,69 @@ private static void checkForRolledLogsAtTime(Calendar cal) throws Exception {
     private static void checkForRolledLogsAtTime(Calendar cal, boolean trace) throws Exception {
-        LOG.logp(Level.INFO, CLASS_NAME, "checkForRolledLogsAtTime", "The next log rollover is scheduled to be at: " + cal.getTime());
+
+        //LOG.logp(Level.INFO, CLASS_NAME, "checkForRolledLogsAtTime", "The next log rollover is scheduled to be at: " + cal.getTime());
```
Remove comment.

riyaaroraa replied:
Done

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -404,33 +424,69 @@ private static void checkForRolledLogsAtTime(Calendar cal) throws Exception {
-            assertTrue("The rolled trace.log file " + traceLog + " doesn't exist.", traceLog.exists());
     }
 
+    //String traceLogName = FILE_SEPARATOR + TRACE_LOG_PREFIX + date + LOG_EXT;
```
Remove comment.

riyaaroraa replied:
Done

pgunapal commented on dev/com.ibm.ws.logging_2_fat/publish/servers/com.ibm.ws.logging.timedrolloverdisabled/server.xml:
```diff
@@ -15,7 +15,7 @@
     
     <featureManager>
-	   <feature>jsp-2.2</feature>
+	   <!-- <feature>jsp-2.2</feature> -->
```
Why is this feature commented out?

riyaaroraa replied:
Done - Commented out during testing for debugging purposes - it has been uncommented and put back

pgunapal reviewed (changes requested):
The fix looks good overall, added some clean up change comments.

riyaaroraa committed:
Address PR review comments for log rollover fix

- Update copyright years to 2022, 2025
- Remove commented out code and cleanup comments
- Remove CWWKO0221E from expected error lists per review
- Uncomment jsp-2.2 feature in server.xml
- Clean up debugging comments and unused setup method

riyaaroraa commented:
fixes #30053

pgunapal reviewed (dismissed):
LGTM

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=21a96731-62bf-4ac9-816d-78e65464576b) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

fmhwong commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/FfdcCleanupTest.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2022, 2023 IBM Corporation and others.
+ * Copyright (c) 2022, 2025 IBM Corporation and others.
```
No need to change the copyright year if there is no change in the file.

riyaaroraa replied:
Done

fmhwong commented on dev/com.ibm.ws.logging_2_fat/publish/servers/com.ibm.ws.logging.timedrolloverxml/server.xml:
```diff
@@ -1,22 +1,10 @@
-<!--
```
Please keep the copyright statement

riyaaroraa replied:
Done

fmhwong replied:
There is no functional change in this file.  Can you revert it to the original?

pgunapal replied:
Revert this file, as no functional changes.

riyaaroraa committed:
Add 30-second tolerance for log rollover timing

- Check exact file first (*.00.0)
- If not found, search for files within same minute
- Handles Mac timing variations (13-30 seconds)
- Fixes SOE build timing issues

Implements request from issue #30053

riyaaroraa committed:
Fix copyright headers per review feedback

- Restore copyright header to server.xml
- Revert copyright year change in FfdcCleanupTest.java

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=0c3cbddb-34b1-45f0-8e53-a84892586409) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

fmhwong commented on dev/com.ibm.ws.logging_2_fat/publish/servers/com.ibm.ws.logging.ffdcCleanup/server.xml:
```diff
@@ -15,7 +15,7 @@
     <include location="../fatTestPorts.xml"/>
     
     <featureManager>
-	   <feature>jsp-2.2</feature>
```
Do we also need to add `timedexit-1.0` here?

riyaaroraa replied:
The timedexit-1.0 feature is included through the chain: this server.xml includes fatTestPorts.xml, which includes fatTestCommon.xml that contains the timedexit-1.0 feature.

pgunapal replied:
The ffdcCleanup server is not being used in this PR's changes, I think we can remove this file as well.

riyaaroraa committed:
Add fatTestPorts.xml to enable timedexit-1.0 feature

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=8bd1f627-c640-4b91-bc23-efb69fc35630) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -404,33 +408,83 @@ private static void checkForRolledLogsAtTime(Calendar cal) throws Exception {
-        if (timeToFirstRollover > 0)
-            Thread.sleep(timeToFirstRollover + FILE_WAIT_SECONDS_PADDING); //sleep until next time the log is set to rollover
+        // If exact file doesn't exist, look for files within the same minute (30-second tolerance)
+        if (!messagesLog.exists()) {
```
This block of code can be modularized into another method, so it can be re-used again, below for traces, instead of using the same code again. Makes the code look cleaner.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -404,33 +408,83 @@ private static void checkForRolledLogsAtTime(Calendar cal) throws Exception {
+        if (trace) {
+            File traceLog = new File(getLogsDirPath(), FILE_SEPARATOR + TRACE_LOG_PREFIX + date + LOG_EXT);
+
+            if (!traceLog.exists()) {
```
Check comment above.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/publish/servers/fatTestPorts.xml:
```diff
@@ -0,0 +1,19 @@
+<!--
```
We don't need to explicitly add this file. This file is already added in automatically, as part of our FAT infrastructure. https://github.com/OpenLiberty/open-liberty/blob/integration/dev/fattest.simplicity/autoFVT-defaults/fatTestPorts.xml

riyaaroraa committed:
Refactor tolerance logic into reusable helper method

- Create findFileWithTolerance helper method
- Eliminate duplicate code for messages and trace files
- Address PR review feedback

riyaaroraa committed:
Implement 30-second tolerance for Mac timing variations

- Add tolerance logic to handle Mac timing delays up to 30 seconds
- Fix edge case in testEnableDisableTimedRollover with 10-second wait
- All 14 tests now passing locally
- Resolves SOE build failures due to timing variations

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -427,24 +425,10 @@ private static void checkForRolledLogsAtTime(Calendar cal, boolean trace) throws
-                    }
-                }
-            }
+            messagesLog = findFileWithTolerance(MESSAGES_LOG_PREFIX, date, LOG_EXT);
```
Consider passing in the `logsDir` into the method and removing the `LOG_EXT`, as it's a global var.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -453,21 +437,9 @@ private static void checkForRolledLogsAtTime(Calendar cal, boolean trace) throws
-                        }
-                    }
-                }
+                traceLog = findFileWithTolerance(TRACE_LOG_PREFIX, date, LOG_EXT);
```
Same comment as above, regarding passing in `logsDir`

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -548,4 +520,22 @@ private String getHttpServlet(String servletPath, LibertyServer server) throws E
     }
+
+    private static File findFileWithTolerance(String logPrefix, String date, String logExt) throws Exception {
+        File logsDir = new File(getLogsDirPath());
```
Don't need to re-get logsDir, if it's being passed.

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -548,4 +520,22 @@ private String getHttpServlet(String servletPath, LibertyServer server) throws E
+        if (files != null) {
+            for (File file : files) {
+                String fileName = file.getName();
+                if (fileName.startsWith(logPrefix) && fileName.endsWith(logExt) &&
```
Consider using the global `LOG_EXT`, instead.

riyaaroraa committed:
Address PR review feedback - optimize helper method and remove unnecessary file

- Pass logsDir as parameter to avoid repeated getLogsDirPath() calls
- Use global LOG_EXT variable instead of parameter
- Cleaner method signature for findFileWithTolerance helper
- Remove fatTestPorts.xml (automatically provided by FAT infrastructure)
- All 14/14 tests passing locally with improved code structure

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -306,9 +310,7 @@ public void testEnableDisableTimedRollover() throws Exception {
-        long nextRollover = cal.getTimeInMillis() - Calendar.getInstance().getTimeInMillis();
-        if (nextRollover > 0)
-            Thread.sleep(nextRollover + FILE_WAIT_SECONDS_PADDING); //sleep for another length of the old specified log rolloverInterval
+        Thread.sleep(10000);
```
Is this sleep for 10 secs needed, why was the above ` long nextRollover = cal.getTimeInMillis() - Calendar.getInstance().getTimeInMillis();` and if condition removed? I think we shouldn't explicitly use sleep here, maybe do a retry instead, if needed?

riyaaroraa committed:
Address PR review feedback - remove unrelated server.xml changes

- Keep only timing tolerance improvements in TimeBasedLogRolloverTest.java
- Remove all server.xml modifications per review feedback

riyaaroraa committed:
Revert all server.xml files to exact original state per review feedback

- Revert timedrolloverxml/server.xml to original copyright and formatting
- Revert ffdcCleanup/server.xml to original whitespace and indentation
- Revert timedrolloverbootstrap/server.xml (remove added timedexit-1.0 feature)
- Keep only timing tolerance improvements in TimeBasedLogRolloverTest.java

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -404,33 +408,57 @@ private static void checkForRolledLogsAtTime(Calendar cal) throws Exception {
         File messagesLog = new File(getLogsDirPath(), messagesLogName);
-        File traceLog = new File(getLogsDirPath(), traceLogName);
 
-        long timeToFirstRollover = cal.getTimeInMillis() - Calendar.getInstance().getTimeInMillis();
```
Does this need to be included back, or is not needed anymore?

riyaaroraa replied:
The timeToFirstRollover line is not needed,  removed the explicit 10 second sleep and kept the original nextRollover logic. The test passes because when rollover is disabled, nextRollover becomes negative, so the if (nextRollover > 0) condition naturally skips the wait. The original logic was sufficient

pgunapal commented on dev/com.ibm.ws.logging_2_fat/fat/src/com/ibm/ws/logging/fat/TimeBasedLogRolloverTest.java:
```diff
@@ -308,7 +312,7 @@ public void testEnableDisableTimedRollover() throws Exception {
         long nextRollover = cal.getTimeInMillis() - Calendar.getInstance().getTimeInMillis();
         if (nextRollover > 0)
-            Thread.sleep(nextRollover + FILE_WAIT_SECONDS_PADDING); //sleep for another length of the old specified log rolloverInterval
+            Thread.sleep(nextRollover + FILE_WAIT_SECONDS_PADDING);
```
Add back the deleted comment.

riyaaroraa replied:
Done

riyaaroraa committed:
Address PR review feedback

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6b026cfb-ebb7-47c8-8619-4fcd2c3be36f) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal reviewed (approved):
LGTM

riyaaroraa commented:
#30053

riyaaroraa commented:
Closes #30053
